plugins/trace_api_plugin/store_provider.cpp

```diff
@@ -166,4 +159,47 @@ namespace eosio::trace_api {
       slice_file.seek(0);
       return true;
    }
+
+
+   void slice_directory::find_or_create_slice_pair(uint32_t slice_number, bool append, fc::cfile& trace, fc::cfile& index) {
+      const bool trace_found = find_or_create_trace_slice(slice_number, append, trace);
+      const bool index_found = find_or_create_index_slice(slice_number, append, index);
+      if (trace_found != index_found) {
+         const std::string trace_status = trace_found ? "existing" : "new";
+         const std::string index_status = index_found ? "existing" : "new";
+         elog("Trace file is ${ts}, but it's metadata file is ${is}. This means the files are not consistent.", ("ts", trace_status)("is", index_status));
+      }
+   }
+
+   void slice_directory::cleanup_old_slices(uint32_t lib) {
+      if (!_minimum_irreversible_history_blocks)
+         return;
+      const uint32_t lib_slice_number = slice_number( lib );
+      if (lib_slice_number < 1 || (_last_cleaned_up_slice && _last_cleaned_up_slice >= lib_slice_number))
+         return;
+
+      // can only cleanup a slice once our last needed history block (lib - *_minimum_irreversible_history_blocks)
+      // is out of that slice (... - width)
+      const int64_t cleanup_block_number = static_cast<int64_t>(lib) - static_cast<int64_t>(*_minimum_irreversible_history_blocks) - _width;
+      if (cleanup_block_number > 0) {
+         uint32_t cleanup_slice_num = slice_number(static_cast<uint32_t>(cleanup_block_number));
+         // since we subtracted width, we are guaranteed cleanup_slice_num is not
+         while (!_last_cleaned_up_slice || *_last_cleaned_up_slice < cleanup_slice_num) {
+            fc::cfile trace;
+            fc::cfile index;
+            const uint32_t slice_to_clean = _last_cleaned_up_slice ? *_last_cleaned_up_slice + 1 : 0;
+            const bool trace_found = find_trace_slice(slice_to_clean, false, trace);
+            if (trace_found) {
+               trace.close();
+               bfs::remove(trace.get_file_path());
+            }
+            const bool index_found = find_index_slice(slice_to_clean, false, index);
```

Comment:
I think we should delete the index and then the data.  I'm not sure if the OS will maintain that order but if it does our reader code will have a better chance at returning a `404` if the index disappears first rather than a temporary `500` because the index was still present but the data was gone.   I realize re-ordering this won't completely close that gap but we should attempt to be consistent here.

Reply:
yep